### PR TITLE
cache babel-register results file by file

### LINKED ISSUE #1769: Improve Babel Cache
We're using babel in our large-size Node.js application, and are facing some issues with the register cache.

The way the cache works now:
* Compilation artifacts are stored in `~/.babel.json`.
* Files that are required are compiled on-the-fly, but only stored in the cache on process exit.

Now, this might work fine for a simple application, but we're facing the following problems:
* Loading `~/.babel.json` take a very long time and takes up a lot of memory:
  
  After running the test cases for our application, the `.babel.json` grows to over 16MB in size. Loading this file on the next run takes a very long time. And the memory usage for this is absurd, as the cache object stays in memory for the whole runtime of the application.

  ```
  > process.memoryUsage();
{ rss: 12742656,
  heapTotal: 7195904,
  heapUsed: 3212688 }
> var cache = require("/home/vagrant/.babel.json");
undefined
> process.memoryUsage();
{ rss: 62210048,
  heapTotal: 57656800,
  heapUsed: 37792696 }
  ```

* Starting up many processes concurrently does not really make use of the cache:

  We run many processes from the same code base, and due to the fact that the cache is only written out once a process exits, starting many processes at the same time means that every process has to compile files on its own and there is no sharing of the compilation artifacts.

I propose the following changes to improve this situation:

Instead of storing all compilation artifacts in a single `.babel.json` file, let's store them in subfolders of a top-level `.babel` folder. We can take the cache-key (including the source `mtime`) that is currently generated, run `sha1` on it and use that as a filename to store the compiled code. Writing out a cache file should happen atomically, and if an error happens during writing of a cache file (because the file was already written by another process), we can just ignore the error.

I hacked up a very basic patch that applies this to babel and ran our applications tests:

Initial run (empty cache):

```
real    4m27.605s
user    2m29.441s
sys     0m46.167s
```

Second run:

```
real    3m40.483s
user    1m44.506s
sys     0m43.362s
```

Here's the times with the old cache:

Initial run (empty cache):

```
real    4m37.994s
user    2m43.912s
sys     0m44.736s
```

Second run:

```
real    3m43.116s
user    1m42.463s
sys     0m46.769s
```

As you can see, the startup numbers do not change for the test runs, but this greatly improves memory usage and concurrent process start up times (which I did not objectively measure yet).

### PR DESCRIPTION
**Pull request by pwmckenna**
Thanks for all the great work! Super grateful to get to use this project. Alright, back to business...

| Q                        | A 
| ------------------------ | ---
| Patch: Bug Fix?          |  no
| Major: Breaking Change?  |  no
| Minor: New Feature?      |  yes?
| Deprecations?            |  no
| Spec Compliancy?         |  n/a
| Tests Added/Pass?        |  no/yes
| Fixed Tickets            | Fixes #1769
| License                  | MIT
| Doc PR                   | no
| Dependency Changes       | ~~[proper-lockfile](https://www.npmjs.com/package/proper-lockfile)~~

babel-register currently stores the entire cache in a single json file, loading it initially then saving it on process exit. This PR caches each compiled file individually, loading it on demand (sync), and saving (async) after compilation.

There are a few "decisions" I made along the way that I should note.
- ~~The individual cache files are locked. This change should help in the case of multiple processes running the same code, but I was worried about file corruption. Not sure if that's a valid concern.~~ *Update:* proper-lockfile doesn't support node 0.12
- The cache file now no longer handles any cache disabling. node.js is responsible for using the cache when appropriate.
- Using md5 to generate unique cache file names, but could use sha1 or really any consistent hash.
- Still keeping the cache in memory as it is populated. Could potentially introduce [lru-cache](https://www.npmjs.com/package/lru-cache) to reduce the memory footprint.

### CODE REVIEW

File: packages/babel-register/src/cache.js [hunk 1]
```diff
@@ -2,53 +2,36 @@ import path from "path";
...
- */
-
-export function load() {
-  if (process.env.BABEL_DISABLE_CACHE) return;
+const cache = {};
 
-  process.on("exit", save);
-  process.nextTick(save);
+export function get(key) {
+  if (!cache.hasOwnProperty(key)) {
```
Comment (xtuc) [inline packages/babel-register/src/cache.js]: You could maybe use this here:
```js
if (!cache[key]) {
```

**Comment by danez** [comment]
Can you describe why you want to change this? Did you have any problems with the current cache behaviour?

**Comment by pwmckenna** [comment]
@danez The main motivation for me is multiple processes that have shared code. The most obvious case is using cluster to spawn multiple copies of the same code, but I even want my mocha tests or webpack build to use cached code, even if I haven't yet killed my server. I also don't necessarily want to load in the cache for all my test files when I'm launching my server, and visa versa.

**Comment by pwmckenna** [comment]
@danez does my explanation make sense? there are no tests for babel-register, so in order to pass the codecov checks I'll likely have to write tests for the entire package...which I don't really want to do unless this is something that would be accepted.

**Comment by xtuc** [comment]
@pwmckenna instead of parsing a big JSON file you want to parse multiple little ones. I don't know if this will increase performance.

**Comment by pwmckenna** [comment]
@xtuc Breaking the cache into multiple small files is just a means to an end. I agree that by itself I wouldn't expect the parsing of json to increase perf (might even hurt reading so many files). I believe that saving the cache immediately after its been updated will have performance benefits when running multiple instances of the process, or running different processes with shared code (background tasks, or even just tests), or if the process doesn't gracefully shutdown. However, saving the entire cache after every file is compiled seems like it would be very slow.

I'd be happy to try to benchmark this stuff if my use case sounds common enough. Ie, when developing having hard shut downs, running tests, or in prod running multiple instances of the process. Are those things worth benchmarking?

**Comment by pwmckenna** [comment]
@danez @xtuc As I sort of figure how how to manual perf test it, another issue I'm running into is the single file means that both my server side and webpack specific compiled versions of the code will be included in the cache file, even though I'm definitely only going to use one. I'll keep testing, but at least in my particular case I'm loading potentially twice as much json as necessary.

**Comment by xtuc** [comment]
@pwmckenna Ok, this is definitely worth benchmarking, thanks for your work :+1:

**Comment by pwmckenna** [comment]
I'm also discovering that doing hot reloading using webpack would benefit quite a bit from caching files within the same process.

Its turning out to be quite difficult to test cache changes within my work project without actually exposing the cache as a configurable option passed into babel-register. perhaps you guys would be open to a pr that made cache configurable, and used the existing one as a default. I've seen in other Issues some objections to adding complexity to babel-register, and I'm not clear what the objection was. Thoughts?

**Comment by pwmckenna** [comment]
scratch that last comment. my lack of caching using babel-loader has nothing to do with babel-register :). Also, I missed the nextTick save that handles saving the cache after the initial burst of require/import statements being compiled. Where this falls short in cases of dynamic require/import statements. I'm curious if simply debouncing the save and calling it after each save might solve for that use case without introducing any additional saves otherwise.

**Comment by pwmckenna** [comment]
Closing this in lieu of https://github.com/babel/babel/pull/5402. Having a hard time convincing myself that my cache benchmarks aren't super specific to my setup, so being able to work on the cache independently sounds appealing. Learned a ton looking into this, though it'll probably just result in README PRs.

**Pull request closed without merging**